File: src/main/java/TextToSpeechHandler.java
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import okhttp3.*;
import org.jetbrains.annotations.NotNull;
import utils.BearerTokenGenerator;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class TextToSpeechHandler implements HttpHandler {

    private static final ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        // 1. CORS 预处理
        Headers respHeaders = exchange.getResponseHeaders();
        respHeaders.add("Access-Control-Allow-Origin", "*");
        respHeaders.add("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
        respHeaders.add("Access-Control-Allow-Headers", "Content-Type, Authorization");

        String method = exchange.getRequestMethod().toUpperCase();
        if ("OPTIONS".equals(method)) {
            exchange.sendResponseHeaders(204, -1);
            return;
        }
        if ("GET".equals(method)) {
            // 返回简单欢迎页
            String html = "<html><head><title>欢迎使用API</title></head><body>"
                    + "<h1>欢迎使用API</h1>"
                    + "<p>此 API 用于与 ChatGPT / Claude 模型交互。您可以发送消息给模型并接收响应。</p>"
                    + "</body></html>";
            byte[] htmlBytes = html.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().add("Content-Type", "text/html; charset=utf-8");
            exchange.sendResponseHeaders(200, htmlBytes.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(htmlBytes);
            }
            return;
        }
        if (!"POST".equals(method)) {
            exchange.sendResponseHeaders(405, -1);
            return;
        }


            try (InputStream clientIs = exchange.getRequestBody()) {
                // 2.1 读取并解析客户端 JSON
                String jsonText = new BufferedReader(new InputStreamReader(clientIs, StandardCharsets.UTF_8))
                        .lines().reduce("", (accumulator, actual) -> accumulator + actual);

                // 用 Jackson 解析 JSON
                Map<String, Object> inJson = objectMapper.readValue(jsonText, Map.class);
                System.out.println("Text To Speech 收到的请求: \n" + objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(inJson));
                String inputText = (String) inJson.get("input");
                String format = inJson.getOrDefault("response_format", "mp3").toString();
                String voice = inJson.getOrDefault("voice", "nova").toString();
                String model = inJson.getOrDefault("model", "tts-1-hd").toString();

                if (!model.startsWith("tts-1")) {
                    model = "tts-1-hd";
                }
                int speed = 1;
                if (inJson.containsKey("speed")) {
                    Object speedObj = inJson.get("speed");
                    if (speedObj instanceof Number) {
                        speed = ((Number) speedObj).intValue();
                    } else {
                        try {
                            speed = Integer.parseInt(speedObj.toString());
                        } catch (Exception ignore) {}
                    }
                }


                Map<String, Object> outJson = new HashMap<>();
                outJson.put("input", inputText);
                outJson.put("response_format", format);
                outJson.put("voice", voice);
                outJson.put("model", model);
                outJson.put("speed", speed);
                if (inJson.containsKey("stream")) {
                    outJson.put("stream", Boolean.parseBoolean(inJson.get("stream").toString()));
                }
                byte[] bodyBytes = objectMapper.writeValueAsBytes(outJson);

                Request ttsRequest = utils.utils.buildRequest(bodyBytes, "/audio/speech");
                OkHttpClient client = utils.utils.getOkHttpClient();

                client.newCall(ttsRequest).enqueue(new Callback() {
                    @Override
                    public void onFailure(@NotNull Call call, @NotNull IOException e) {
                        e.printStackTrace();
                    }

                    @Override
                    public void onResponse(@NotNull Call call, @NotNull Response ttsResp) {
                        if (!ttsResp.isSuccessful() || ttsResp.body() == null) {
                            System.err.println("TTS 服务返回错误，code=" + ttsResp.code());
                            return;
                        }

                        // 设置响应头，使用 chunked 传输
                        Headers hdrs = exchange.getResponseHeaders();
                        hdrs.add("Content-Type", "audio/mpeg");
                        hdrs.add("Accept-Ranges", "bytes");
                        hdrs.add("Content-Disposition", "attachment; filename=\"output.mp3\"");
                        try {
                            // length=0 即启用 chunked 模式
                            exchange.sendResponseHeaders(200, 0);
                            InputStream ttsStream = ttsResp.body().byteStream();
                            OutputStream clientOs = exchange.getResponseBody();

                            // 使用较小的缓冲区以减少延迟
                            byte[] buffer = new byte[1024];
                            int len;
                            while ((len = ttsStream.read(buffer)) != -1) {
                                clientOs.write(buffer, 0, len);
                                clientOs.flush(); // 立即刷新，确保数据尽快发送到客户端
                            }
                        } catch (IOException ioe) {
                            ioe.printStackTrace();
                        } finally {
                            try {
                                exchange.getResponseBody().close();
                            } catch (IOException ignored) {}
                        }
                    }

                });

            } catch (Exception ex) {
                ex.printStackTrace();
            }
    }
}

File: src/main/cpp/text_to_speech_handler.cpp
#include "httplib.h"
#include "text_to_speech_handler.h"
#include "bearer_token_generator.h"
#include <curl/curl.h>
#include <chrono>
#include <ctime>
#include <sstream>
#include <iomanip>
#include <vector>
#include <cstring>
#include <zlib.h>

// Helper: Write callback for libcurl
static size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp) {
    size_t totalSize = size * nmemb;
    std::vector<char>* buf = static_cast<std::vector<char>*>(userp);
    buf->insert(buf->end(), (char*)contents, (char*)contents + totalSize);
    return totalSize;
}

// Helper: Gzip decompress
static std::string decompressGzip(const std::vector<char>& compressed) {
    if (compressed.size() < 4) return "";
    z_stream zs{};
    zs.next_in = (Bytef*)compressed.data();
    zs.avail_in = compressed.size();

    if (inflateInit2(&zs, 16 + MAX_WBITS) != Z_OK) return "";

    std::string out;
    char buffer[8192];
    int ret;
    do {
        zs.next_out = reinterpret_cast<Bytef*>(buffer);
        zs.avail_out = sizeof(buffer);
        ret = inflate(&zs, Z_NO_FLUSH);
        if (out.size() < zs.total_out)
            out.append(buffer, zs.total_out - out.size());
    } while (ret == Z_OK);
    inflateEnd(&zs);
    return (ret == Z_STREAM_END) ? out : "";
}

void TextToSpeechHandler::handle(const httplib::Request& req, httplib::Response& res) {
    // 1. CORS Preprocessing
    res.set_header("Access-Control-Allow-Origin", "*");
    res.set_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
    res.set_header("Access-Control-Allow-Headers", "Content-Type, Authorization");

    if (req.method == "OPTIONS") {
        res.status = 204;
        return;
    }

    if (req.method == "GET") {
        res.set_content("<html><head><title>欢迎使用API</title></head><body>"
                        "<h1>欢迎使用API</h1>"
                        "<p>此 API 用于与 ChatGPT / Claude 模型交互。您可以发送消息给模型并接收响应。</p>"
                        "</body></html>", "text/html; charset=utf-8");
        res.status = 200;
        return;
    }

    if (req.method != "POST") {
        res.status = 405;
        return;
    }

    // 2. Parse client JSON
    try {
        auto inJson = json::parse(req.body);
        std::string inputText = inJson.at("input").get<std::string>();
        std::string format = inJson.value("response_format", "mp3");
        std::string voice = inJson.value("voice", "nova");
        std::string model = inJson.value("model", "tts-1-hd");
        int speed = inJson.value("speed", 1);
        if (!model.starts_with("tts-1")) model = "tts-1-hd";

        json outJson = {
            {"input", inputText},
            {"response_format", format},
            {"voice", voice},
            {"model", model},
            {"speed", speed}
        };
        if (inJson.contains("stream")) {
            outJson["stream"] = inJson["stream"];
        }
        std::string bodyStr = outJson.dump();

        // 3. Prepare headers
        auto now = std::chrono::system_clock::now();
        auto itt = std::chrono::system_clock::to_time_t(now);
        std::ostringstream ss;
        ss << std::put_time(std::gmtime(&itt), "%Y-%m-%dT%H:%M:%SZ");
        std::string formattedDate = ss.str();

        std::vector<unsigned char> bodyContent(bodyStr.begin(), bodyStr.end());
        std::string path = "/audio/speech";
        std::string bearer = BearerTokenGenerator::GetBearer(bodyContent, path, formattedDate, "POST");

        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Content-Type: application/json; charset=UTF-8");
        headers = curl_slist_append(headers, "Accept: application/json");
        headers = curl_slist_append(headers, ("Date: " + formattedDate).c_str());
        headers = curl_slist_append(headers, "Client-time-zone: -04:00");
        headers = curl_slist_append(headers, ("Authorization: " + bearer).c_str());
        headers = curl_slist_append(headers, ("User-Agent: " + BearerTokenGenerator::UA).c_str());
        headers = curl_slist_append(headers, "Accept-Language: en-US");
        headers = curl_slist_append(headers, "X-Cl-Options: hb");
        headers = curl_slist_append(headers, "Accept-Encoding: gzip");
        headers = curl_slist_append(headers, "Connection: Keep-Alive");

        // 4. Send request to TTS API
        CURL* curl = curl_easy_init();
        if (!curl) {
            res.status = 500;
            return;
        }
        curl_easy_setopt(curl, CURLOPT_URL, "https://api.chaton.ai/audio/speech");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, bodyStr.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, bodyStr.size());

        std::vector<char> responseBuf;
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &responseBuf);

        // Accept compressed response
        curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "gzip");

        CURLcode code = curl_easy_perform(curl);
        long http_code = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);

        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);

        if (code != CURLE_OK || http_code != 200) {
            res.status = 502;
            return;
        }

        // 5. Check for gzip and decompress if needed
        std::string audioData;
        // Try decompressing, fallback to raw if not gzip
        audioData = decompressGzip(responseBuf);
        if (audioData.empty()) {
            audioData.assign(responseBuf.begin(), responseBuf.end());
        }

        // 6. Set response headers and stream audio
        res.set_header("Content-Type", "audio/mpeg");
        res.set_header("Accept-Ranges", "bytes");
        res.set_header("Content-Disposition", "attachment; filename=\"output.mp3\"");
        res.status = 200;
        res.body = std::move(audioData);

    } catch (const std::exception& ex) {
        res.status = 500;
    }
}
